typed/merge.go

```diff
@@ -17,11 +17,62 @@ limitations under the License.
 package typed
 
 import (
+	"sort"
+
 	"sigs.k8s.io/structured-merge-diff/fieldpath"
 	"sigs.k8s.io/structured-merge-diff/schema"
 	"sigs.k8s.io/structured-merge-diff/value"
 )
 
+type pathElementValueMap struct {
+	members sortedPathElementValues
+}
+
+type pathElementValue struct {
+	PathElement fieldpath.PathElement
+	Value       value.Value
+}
+
+type sortedPathElementValues []pathElementValue
+
+// Implement the sort interface; this would permit bulk creation, which would
+// be faster than doing it one at a time via Insert.
+func (spev sortedPathElementValues) Len() int { return len(spev) }
+func (spev sortedPathElementValues) Less(i, j int) bool {
+	return spev[i].PathElement.Less(spev[j].PathElement)
+}
+func (spev sortedPathElementValues) Swap(i, j int) { spev[i], spev[j] = spev[j], spev[i] }
+
+// Insert adds pe to the set.
+func (s *pathElementValueMap) Insert(pe fieldpath.PathElement, v value.Value) {
+	loc := sort.Search(len(s.members), func(i int) bool {
+		return !s.members[i].PathElement.Less(pe)
+	})
+	if loc == len(s.members) {
+		s.members = append(s.members, pathElementValue{pe, v})
+		return
+	}
+	if s.members[loc].PathElement.Equals(pe) {
+		return
+	}
+	s.members = append(s.members, pathElementValue{})
+	copy(s.members[loc+1:], s.members[loc:])
+	s.members[loc] = pathElementValue{pe, v}
+}
+
+func (s *pathElementValueMap) Get(pe fieldpath.PathElement) *value.Value {
+	loc := sort.Search(len(s.members), func(i int) bool {
+		return !s.members[i].PathElement.Less(pe)
+	})
+	if loc == len(s.members) {
+		return nil
+	}
+	if s.members[loc].PathElement.Equals(pe) {
+		return &s.members[loc].Value
```

Comment:
This is very risky, a subsequent insert will make the pointer point to the wrong thing.

Reply:
yeah, we could make it return a copy, bool to be safer?

Reply:
Yes, that'd be better.